File: packages/tauri_plugin_jarvis/webview-src/src/api/system.ts
import { invoke } from "@tauri-apps/api";

export function openTrash() {
  return invoke("plugin:jarvis|open_trash");
}

export function emptyTrash() {
  return invoke("plugin:jarvis|empty_trash");
}

export function shutdown() {
  return invoke("plugin:jarvis|shutdown");
}

export function reboot() {
  return invoke("plugin:jarvis|reboot");
}

export function sleep() {
  return invoke("plugin:jarvis|sleep");
}

export function lock() {
  return invoke("plugin:jarvis|lock");
}

export function toggleSystemAppearance() {
  return invoke("plugin:jarvis|toggle_system_appearance");
}

export function showDesktop() {
  return invoke("plugin:jarvis|show_desktop");
}

export function quitAppApps() {
  return invoke("plugin:jarvis|quit_app_apps");
}

export function sleepDisplays() {
  return invoke("plugin:jarvis|sleep_displays");
}

export function setVolume() {
  return invoke("plugin:jarvis|set_volume");
}

export function turnVolumeUp() {
  return invoke("plugin:jarvis|turn_volume_up");
}

export function turnVolumeDown() {
  return invoke("plugin:jarvis|turn_volume_down");
}

export function toggleStageManager() {
  return invoke("plugin:jarvis|toggle_stage_manager");
}

export function toggleBluetooth() {
  return invoke("plugin:jarvis|toggle_bluetooth");
}

export function toggleHiddenFiles() {
  return invoke("plugin:jarvis|toggle_hidden_files");
}

export function ejectAllDisks() {
  return invoke("plugin:jarvis|eject_all_disks");
}

export function logoutUser() {
  return invoke("plugin:jarvis|logout_user");
}

export function toggleMute() {
  return invoke("plugin:jarvis|toggle_mute");
}

export function mute() {
  return invoke("plugin:jarvis|mute");
}

export function unmute() {
  return invoke("plugin:jarvis|unmute");
}

export function hideAllAppsExceptFrontmost() {
  return invoke("plugin:jarvis|hide_all_apps_except_frontmost");
}

File: apps/desktop/src-tauri/src/commands/system.rs
use super::utils::run_apple_script;

#[tauri::command]
pub async fn open_trash() -> Result<(), String> {
    // #[cfg(target_os = "macos")]
    run_apple_script("tell application \"Finder\" to open trash")
    // #[cfg(target_os = "linux")]
    // run_linux_command("xdg-open trash")
}

#[tauri::command]
pub async fn empty_trash() -> Result<(), String> {
    run_apple_script("tell application \"Finder\" to empty the trash")
}

#[tauri::command]
pub async fn shutdown() -> Result<(), String> {
    run_apple_script("tell application \"System Events\" to shut down")
}

#[tauri::command]
pub async fn reboot() -> Result<(), String> {
    run_apple_script("tell application \"System Events\" to restart")
}

#[tauri::command]
pub async fn sleep() -> Result<(), String> {
    run_apple_script("tell application \"System Events\" to sleep")
}

/// ```applescript
/// tell application "System Events" to tell appearance preferences
///   set dark_mode to dark mode
///   set dark_mode to not dark_mode
///   set dark mode to dark_mode
/// end tell
/// ```
#[tauri::command]
pub async fn toggle_system_appearance() -> Result<(), String> {
    run_apple_script(
        r#"
        tell application "System Events" to tell appearance preferences
            set dark_mode to dark mode
            set dark_mode to not dark_mode
            set dark mode to dark_mode
        end tell
    "#,
    )
}

#[tauri::command]
pub async fn show_desktop() -> Result<(), String> {
    run_apple_script("tell application \"System Events\" to key code 103")
}

#[tauri::command]
pub async fn quit_all_apps() -> Result<(), String> {
    todo!()
}

#[tauri::command]
pub async fn sleep_displays() -> Result<(), String> {
    run_apple_script("do shell script \"pmset displaysleepnow\"")
}

/// Set Volume to 0
#[tauri::command]
pub async fn set_volume(percentage: u8) -> Result<(), String> {
    run_apple_script(&format!("set volume output volume {}", percentage))
}
/// Turn Volume Up
#[tauri::command]
pub async fn turn_volume_up() -> Result<(), String> {
    run_apple_script("set volume output volume (output volume of (get volume settings) + 10)")
}
/// Turn Volume Down
#[tauri::command]
pub async fn turn_volume_down() -> Result<(), String> {
    run_apple_script("set volume output volume (output volume of (get volume settings) - 10)")
}
/// Toggle Stage Manager
#[tauri::command]
pub async fn toggle_stage_manager() -> Result<(), String> {
    todo!()
}

/// Toggle Bluetooth
#[tauri::command]
pub async fn toggle_bluetooth() -> Result<(), String> {
    todo!()
}

/// Toggle Hidden Files
#[tauri::command]
pub async fn toggle_hidden_files() -> Result<(), String> {
    todo!()
}

/// Eject All Disks
#[tauri::command]
pub async fn eject_all_disks() -> Result<(), String> {
    run_apple_script("tell application \"Finder\" to eject (every disk whose ejectable is true)")
}
/// Logout <User>
#[tauri::command]
pub async fn logout_user() -> Result<(), String> {
    run_apple_script("tell application \"System Events\" to log out")
}
/// Toggle Mute
#[tauri::command]
pub async fn toggle_mute() -> Result<(), String> {
    run_apple_script("set volume output muted not (output muted of (get volume settings))")
}

#[tauri::command]
pub async fn mute() -> Result<(), String> {
    run_apple_script("set volume with output muted")
}

#[tauri::command]
pub async fn unmute() -> Result<(), String> {
    run_apple_script("set volume without output muted")
}
/// Hide All Apps Except Frontmost
#[tauri::command]
pub async fn hide_all_apps_except_frontmost() -> Result<(), String> {
    run_apple_script(
        r#"tell application "System Events"
            set frontApp to name of first application process whose frontmost is true
            set visibleApps to every process whose visible is true and name is not frontApp
            repeat with theApp in visibleApps
                set visible of theApp to false
            end repeat
        end tell"#,
    )
}
